tests/gem/mock_gem.py

```diff
@@ -0,0 +1,30 @@
+import asyncio
+from typing import List, Optional, Tuple
+
+
+class MockGemProtocol(asyncio.Protocol):
+    def __init__(self) -> None:
+        self._transport: Optional[asyncio.WriteTransport] = None
+        self._expects_and_replies: List[Tuple[bytes, bytes]] = []
+        self._buffer = bytearray()
+
+    def connection_made(self, transport: asyncio.WriteTransport) -> None:
+        self._transport = transport
+
+    def data_received(self, data: bytes) -> None:
+        assert self._transport is not None
+
+        self._buffer += data
+
+        while len(self._buffer) > 0 and len(self._expects_and_replies) > 0:
+            expected, reply = self._expects_and_replies[0]
+            if self._buffer.startswith(expected):
+                del self._buffer[0 : len(expected)]
+                self._transport.write(reply)
+                del self._expects_and_replies[0]
+            else:
+                # Should have received the first part of it anyway
+                assert expected.startswith(self._buffer)
```

Comment:
It looks like `pytest` has [some helpers](https://docs.pytest.org/en/6.2.x/example/simple.html#writing-well-integrated-assertion-helpers) that will play well with the output if the test fails that we probably want to use here.

Reply:
I'm a little confused about how the testing fits together. They're all written with `unittest` but run with `pytest`; can `pytest` helpers be used?

Reply:
Seems to work fine